probe/proc/reader_linux.go

```diff
@@ -0,0 +1,136 @@
+package proc
+
+import (
+	"bytes"
+	"path"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+var tcpFiles = []string{
+	"net/tcp",
+	"net/tcp6",
+}
+
+// the Linux "/proc" reader
+type reader struct {
+	proc      Dir
+	handles   *filesCache
+	withProcs bool
+}
+
+// NewReader creates a new /proc reader.
+func NewReader(proc Dir, withProcs bool) Reader {
+	return &reader{
+		proc:      proc,
+		handles:   newFilesCache(proc),
+		withProcs: withProcs,
+	}
+}
+
+// Close closes the Linux "/proc" reader
+func (r *reader) Close() error {
+	return r.handles.Close()
+}
+
+// Processes walks the /proc directory and marshalls the files into
+// instances of Process, which it then passes one-by-one to the
+// supplied function. Processes() is only made public so that is
+// can be tested.
+func (r *reader) Processes(f func(Process)) error {
+	dirEntries, err := r.proc.ReadDirNames(r.proc.Root())
+	if err != nil {
+		return err
+	}
+
+	var fdStat syscall.Stat_t
+	buf := bytes.Buffer{}
+
+	for _, subdir := range dirEntries {
+		readIntoBuffer := func(filename string, pinned bool) error {
+			buf.Reset()
+			return r.handles.ReadInto(path.Join(subdir, filename), &buf, pinned)
+		}
+
+		pid, err := strconv.Atoi(subdir)
+		if err != nil {
+			continue
+		}
+
+		if readIntoBuffer("stat", false) != nil {
+			continue
+		}
+		splits := strings.Fields(buf.String())
+		ppid, err := strconv.Atoi(splits[3])
+		if err != nil {
+			return err
+		}
+
+		threads, err := strconv.Atoi(splits[19])
+		if err != nil {
+			return err
+		}
+
+		cmdline := ""
+		if readIntoBuffer("cmdline", true) == nil {
+			cmdlineBuf := bytes.Replace(buf.Bytes(), []byte{'\000'}, []byte{' '}, -1)
+			cmdline = string(cmdlineBuf)
+		}
+
+		comm := "(unknown)"
+		if readIntoBuffer("comm", true) == nil {
+			comm = strings.TrimSpace(buf.String())
+		}
+
+		fdBase := path.Join(r.proc.Root(), strconv.Itoa(pid), "fd")
+		fdNames, err := r.proc.ReadDirNames(fdBase)
+		if err != nil {
+			return err
+		}
+
+		inodes := []uint64{}
+		for _, fdName := range fdNames {
+			// Direct use of syscall.Stat() to save garbage.
+			fdPath := path.Join(fdBase, fdName)
+			err = syscall.Stat(fdPath, &fdStat)
+			if err == nil && (fdStat.Mode&syscall.S_IFMT == syscall.S_IFSOCK) { // We want sockets only.
+				inodes = append(inodes, fdStat.Ino)
+			}
+		}
+
+		f(Process{
+			PID:     pid,
+			PPID:    ppid,
+			Comm:    comm,
+			Cmdline: cmdline,
+			Threads: threads,
+			Inodes:  inodes,
+		})
+	}
+
+	return nil
+}
+
+// Connections walks through all the connections in the "/proc"
+// Note: we never perform the process<->connection matching in Linux:
+//       use the CachingReader instead
+func (r *reader) Connections(f func(Connection)) error {
+	buf := bytes.Buffer{}
+	for _, tcpFile := range tcpFiles {
```

Comment:
Is this no longer reading the /proc/<pid>/net/tcp files?  Or have I misunderstood?